README.md
# attack-flow-whodunit
advanced topics work 

this repository contains all the work from Advanced Topics research project

**Scalable Metagraph Algorithms for Provable Network Security**

## attackflows
folder contains `.afb` and `.json` files of 40 attacks used as the database of this research project

## whodunit
folder contains `whodunit` code from https://gitlab.com/bontchev/whodunit with modified version `whodunit-Hattie.py` that is able to:
* output groups with high matching percentage
* output groups with a low number of missing techniques
* list the missing techniques
* output position of missing techniques in attack sequence of each threat actor
* output groups with low missing score

***run the code:***
Use the command 

```
python whodunit-Hattie.py [-h] [-v] [-n NUMGROUPS] [-b] [-u] [-d] FILE_PATH_OF_TECHNIQUES_LIST
```

* input file must be of extension `.txt` only
* text file contains list of techniques used in attack


whodunit/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.3]

### Added in version 1.0.3

* The program can now process more than a single incident report file.

### Changed in version 1.0.3

* Minor fixes to the documentation.

## [1.0.2]

### Changed in version 1.0.2

* Had forgotten to bump the version number in 1.0.1. Fixed.

* Better message syntax when reporting only a single group.

* The `-u` option can now be used on its own (i.e., no `techniques` argument
is needed then).

## [1.0.1]

### Added in version 1.0.1

* Implemented the `-d` option for support of deprecated techniques.

### Changed in version 1.0.1

* Updated the documentation to note that only the `Enterprise` matrix of the
MITRE ATT&CK framework is supported (and not the `Mobile` or `ICS` matrices).

* Minor language fixes in `README.md` and `report.txt`.

* Simplified the script a bit by not storing the set of common attacks and
just keeping the confidence number.

* Fixed a bug where a technique listed in the incident report would not be
recognized as valid, if it was not used by a known group.

* Re-arranged and optimized the code, in order to reduce the number of times
that the MITRE ATT&CK structure is traversed for groups and techniques.

## [1.0.0]

### Added in version 1.0.0

* Initial release


whodunit/whodunit.py
#!/usr/bin/env python


from __future__ import print_function


__description__ = "Recognize an APT group from the techniques used"
__license__ = "GPL"
__VERSION__ = "1.0.3"
__uri__ = "https://gitlab.com/bontchev/whodunit"
__author__ = "Vesselin Bontchev"
__email__ = "vbontchev@yahoo.com"


from argparse import ArgumentParser
from heapq import nlargest
from os import access, R_OK
from os.path import isfile
from sys import argv, stderr, version_info


if version_info < (3, 6):
    print("Python 3.6 or higher required.", file=stderr)
    exit()

try:
    from enterpriseattack import Attack
except ImportError:
    print(
        'Could not import module "enterpriseattack"; try "pip install enterpriseattack".',
        file=stderr,
    )
    exit()


def get_options():
    parser = ArgumentParser(description=__description__)
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        version="%(prog)s version {}".format(__VERSION__),
    )
    parser.add_argument(
        "-n",
        "--numgroups",
        type=int,
        default=10,
        help="Number of closest groups to list (default: %(default)s)",
    )
    parser.add_argument(
        "-b", "--verbose", action="store_true", help="Verbose operation"
    )
    parser.add_argument(
        "-u", "--update", action="store_true", help="Update the MITRE ATT&CK data"
    )
    parser.add_argument(
        "-d",
        "--deprecated",
        action="store_true",
        help="Don't ignore the deprecated techniques",
    )
    parser.add_argument(
        "techniques",
        nargs="*" if "-u" in argv or "--update" in argv else "+",
        help="File containing a list of observed techniques",
    )
    return parser.parse_args()


def read_report(report_file, all_techniques, verbose):
    print('Processing report file "{}"...'.format(report_file))
    with open(report_file, encoding="utf-8") as f:
        data = f.readlines()

    techniques = set()
    for line in data:
        line = line.strip()
        pos = line.find("#")
        if pos >= 0:
            line = line[:pos].strip()
        tids = line.split()
        for tid in tids:
            tid = tid.upper()
            if tid in all_techniques:
                techniques.add(tid)
            else:
                print('Unrecognized technique: "{}".'.format(tid), file=stderr)
                exit()

    if verbose:
        print("Observed techniques:")
        for technique in sorted(techniques):
            print(technique)
        print()

    return techniques


def get_groups_and_techniques(attack, techniques):
    def get_techniques(groupObj):
        mergedTechsAndSubs = groupObj.techniques + groupObj.sub_techniques
        return [t.id for t in mergedTechsAndSubs]

    groups = []
    max_name = 0
    for group in attack.groups:
        the_group = {}
        the_group["name"] = group.name
        if len(group.name) > max_name:
            max_name = len(group.name)
        the_group["id"] = group.id
        the_techniques = set(get_techniques(group))
        the_group["techniques"] = the_techniques
        common_techniques = the_techniques.intersection(techniques)
        the_group["confidence"] = len(common_techniques) * 100.0 / len(techniques)
        groups.append(the_group)
    return groups, max_name


def match_groups(groups, num_groups, max_name):
    closest_groups = nlargest(num_groups, groups, key=lambda x: x["confidence"])
    if num_groups > 1:
        print("The {} APT groups".format(num_groups), end="")
    else:
        print("The APT group", end="")
    print(" most likely responsible:")
    for group in sorted(closest_groups, key=lambda x: (-x["confidence"], x["id"])):
        print(
            "{} ({:<{width}}): {:>6.2f}%".format(
                group["id"], group["name"], group["confidence"], width=max_name
            )
        )


def main():
    args = get_options()
    attack = Attack(include_deprecated=args.deprecated, update=args.update)
    if len(args.techniques) == 0:
        exit()
    multi_report = len(args.techniques) > 1
    all_techniques = [
        x.id for x in list(attack.techniques) + list(attack.sub_techniques)
    ]
    for report_file in args.techniques:
        if isfile(report_file) and access(report_file, R_OK):
            techniques = read_report(report_file, all_techniques, args.verbose)
            groups, max_name = get_groups_and_techniques(attack, techniques)
            match_groups(groups, args.numgroups, max_name)
            if multi_report:
                print()
        else:
            print('Could not open file "{}"'.format(report_file), file=stderr)


if __name__ == "__main__":
    main()


ml-package/document_analysis_Hattie.py
from cmath import pi
import pickle
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer, porter
import numpy as np
from utils.filter_data import *
import pandas as pd
import re
from nltk.tokenize import sent_tokenize
from utils.csv_output import Classifier_results, CSVOutput
import os
import csv
import requests
from tika import parser
import sys
from datetime import datetime

from deepl_utils import *

techniques_df = pd.read_csv("dataset_new.csv")

def repl(matchobj):
    return ","+ matchobj.group(1) + ","

def remove_empty_lines(text):
	lines = text.split("\n")
	non_empty_lines = [line for line in lines if line.strip() != ""]

	string_without_empty_lines = ""
	for line in non_empty_lines:
		if line != "\n": 
			string_without_empty_lines += line + "\n"

	return string_without_empty_lines 

def combine_text(list_of_text):
    combined_text = ' '.join(list_of_text)
    return combined_text

def lemmatize_set(dataset):
    lemmatizer = WordNetLemmatizer()
    lemmatized_list = []
    for sentence in dataset:
        word_list = word_tokenize(sentence)
        lemma_list = [lemmatizer.lemmatize(w) for w in word_list]
        lemmatized_list.append(' '.join(lemma_list))
    return lemmatized_list

def stemmatize_set(dataset):
    ps = porter.PorterStemmer()
    stemmatize_list = []
    for sentence in dataset:
        word_list = word_tokenize(sentence)
        stemma_list = [ps.stem(w) for w in word_list]
        stemmatize_list.append(' '.join(stemma_list))
    return stemmatize_list

def f_measure(recall, precision):
    if recall != 0 and precision != 0:
        return (2*precision*recall)/(precision+recall)
    else:
        return 0.01

def extract_text(file_path):

    parsed_document = parser.from_file(file_path)
    lines = []
    lines.append(parsed_document['content'])
    # print(lines)
    return lines


ml_model_filenames = ['ml_models/MLP_classifier.sav', 'ml_models/Logreg.sav', 'ml_models/Multinomial_NB.sav', 'ml_models/SVM_Classifier_OVR.sav']
                    #, 'SVM_Classifier_OVO.sav', 'Logreg_normale.sav']

def analyze_all_doc(file_path, model_filenames):

    lines = extract_text(file_path)

    
    ## Apply regex 
    regex_list = load_regex("utils/regex.yml")

    text = combine_text(lines)
    text = re.sub('(%(\w+)%(\/[^\s]+))', repl, text)
    text = apply_regex_to_string(regex_list, text)
    text = re.sub('\(.*?\)', '', text)
    text = remove_empty_lines(text)
    text = text.strip()
    sentences = sent_tokenize(text)

    # print(len(sentences))

    num_sen = len(sentences)

    double_sentences = []

    for i in range(1, len(sentences)):
        new_sen = sentences[i-1] + sentences[i]
        double_sentences.append(new_sen)

    results = []
    
    for model_filename in model_filenames: 
        # load the model from disk
        vectorizer, classifier = pickle.load(open(model_filename, 'rb'))

        stemmatized_set = stemmatize_set(sentences)
        lemmatized_set = lemmatize_set(stemmatized_set)
        x_test_vectors = vectorizer.transform(lemmatized_set)
        predicted = classifier.predict(x_test_vectors)
        #Matrix
        #Vector of vector of probabilities
        predict_proba_scores = classifier.predict_proba(x_test_vectors)
        #Identify the indexes of the top predictions (increasing order so let's take the last 2, highest proba)
        top_k_predictions = np.argsort(predict_proba_scores, axis = 1)[:,-2:]
        #Get classes related to previous indexes
        top_class_v = classifier.classes_[top_k_predictions]

        results = {'ID': [], 'Name': [], 'Sentence': [],}
        df = pd.DataFrame(results)

        thresholds = [0.10, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]
        for threshold in thresholds: 
            # accepted = []

            for i in range(0,len(predict_proba_scores)):
                sorted_indexes = top_k_predictions[i]
                top_classes = top_class_v[i]
                proba_vector = predict_proba_scores[i]
                predicted_labels = []
                label_name = []
                used_sentences = []
                if proba_vector[sorted_indexes[1]] > threshold:
                    # accepted.append(top_classes[1])
                    predicted_label = top_classes[1]
                    label_name = " ".join(re.findall("[a-zA-Z]+", techniques_df[techniques_df["label_subtec"] == predicted_label]["tec_name"].head(1).to_string()))
                    predicted_labels.append(predicted_label)
                    used_sentences.append(sentences[i])
                    result = {"ID": predicted_label, "Name": label_name, "Sentence": sentences[i],}
                    
                    df.loc[len(df)] = result
    df.drop_duplicates(inplace=True)
    return df

# total arguments
n = len(sys.argv)
print("Start analysing", n-1, "documents...")

for i in range(1, n):
    file_path = sys.argv[i]
    results = analyze_all_doc(file_path, ml_model_filenames)

    index = 0
    for char in file_path:
        if char == '.':
            break
        index+=1

    new_file_path = file_path[:index] + '_result.xlsx'
    # print(new_file_path)
    results.to_excel(new_file_path, index=False)
    print("Analysis progress: ", i, "/", n-1, sep="")

    # convert results xlsx to csv 
    technique_list = pd.read_excel(new_file_path, 'Sheet1')
    csv_file_path = file_path[:index] + '_result.csv'
    technique_list.to_csv(csv_file_path)
    print("Technique list result is successfully converted to csv file.")

    # filter csv to extract technique IDs only
    filtered_csv = pd.read_csv(csv_file_path, usecols=['ID'])
    txt_file_path = file_path[:index] + '_result.txt'
    filtered_csv.to_csv(txt_file_path, header=False, index=False)
    print("List of Technique ID is successfully extracted to txt file.")

ml-package/document_analysis_modified_v1.py
from cmath import pi
import pickle
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer, porter
import numpy as np
from utils.filter_data import *
import pandas as pd
import re
from nltk.tokenize import sent_tokenize
from utils.csv_output import Classifier_results, CSVOutput
import os
import csv
import requests
from tika import parser
import sys
from datetime import datetime

from deepl_utils import *

techniques_df = pd.read_csv("dataset_new.csv")

def repl(matchobj):
    return ","+ matchobj.group(1) + ","

def remove_empty_lines(text):
	lines = text.split("\n")
	non_empty_lines = [line for line in lines if line.strip() != ""]

	string_without_empty_lines = ""
	for line in non_empty_lines:
		if line != "\n": 
			string_without_empty_lines += line + "\n"

	return string_without_empty_lines 

def combine_text(list_of_text):
    combined_text = ' '.join(list_of_text)
    return combined_text

def lemmatize_set(dataset):
    lemmatizer = WordNetLemmatizer()
    lemmatized_list = []
    for sentence in dataset:
        word_list = word_tokenize(sentence)
        lemma_list = [lemmatizer.lemmatize(w) for w in word_list]
        lemmatized_list.append(' '.join(lemma_list))
    return lemmatized_list

def stemmatize_set(dataset):
    ps = porter.PorterStemmer()
    stemmatize_list = []
    for sentence in dataset:
        word_list = word_tokenize(sentence)
        stemma_list = [ps.stem(w) for w in word_list]
        stemmatize_list.append(' '.join(stemma_list))
    return stemmatize_list

def f_measure(recall, precision):
    if recall != 0 and precision != 0:
        return (2*precision*recall)/(precision+recall)
    else:
        return 0.01

def extract_text(file_path):

    parsed_document = parser.from_file(file_path)
    lines = []
    lines.append(parsed_document['content'])
    # print(lines)
    return lines


ml_model_filenames = ['ml_models/MLP_classifier.sav', 'ml_models/Logreg.sav', 'ml_models/Multinomial_NB.sav', 'ml_models/SVM_Classifier_OVR.sav']
                    #, 'SVM_Classifier_OVO.sav', 'Logreg_normale.sav']

def analyze_all_doc(file_path, model_filenames):

    lines = extract_text(file_path)

    
    ## Apply regex 
    regex_list = load_regex("utils/regex.yml")

    text = combine_text(lines)
    text = re.sub('(%(\w+)%(\/[^\s]+))', repl, text)
    text = apply_regex_to_string(regex_list, text)
    text = re.sub('\(.*?\)', '', text)
    text = remove_empty_lines(text)
    text = text.strip()
    sentences = sent_tokenize(text)

    # print(len(sentences))

    num_sen = len(sentences)

    double_sentences = []

    for i in range(1, len(sentences)):
        new_sen = sentences[i-1] + sentences[i]
        double_sentences.append(new_sen)

    results = []
    
    for model_filename in model_filenames: 
        # load the model from disk
        vectorizer, classifier = pickle.load(open(model_filename, 'rb'))

        stemmatized_set = stemmatize_set(sentences)
        lemmatized_set = lemmatize_set(stemmatized_set)
        x_test_vectors = vectorizer.transform(lemmatized_set)
        predicted = classifier.predict(x_test_vectors)
        #Matrix
        #Vector of vector of probabilities
        predict_proba_scores = classifier.predict_proba(x_test_vectors)
        #Identify the indexes of the top predictions (increasing order so let's take the last 2, highest proba)
        top_k_predictions = np.argsort(predict_proba_scores, axis = 1)[:,-2:]
        #Get classes related to previous indexes
        top_class_v = classifier.classes_[top_k_predictions]

        results = {'ID': [], 'Name': [], 'Sentence': [],}
        df = pd.DataFrame(results)

        thresholds = [0.10, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]
        for threshold in thresholds: 
            # accepted = []

            for i in range(0,len(predict_proba_scores)):
                sorted_indexes = top_k_predictions[i]
                top_classes = top_class_v[i]
                proba_vector = predict_proba_scores[i]
                predicted_labels = []
                label_name = []
                used_sentences = []
                if proba_vector[sorted_indexes[1]] > threshold:
                    # accepted.append(top_classes[1])
                    predicted_label = top_classes[1]
                    label_name = " ".join(re.findall("[a-zA-Z]+", techniques_df[techniques_df["label_subtec"] == predicted_label]["tec_name"].head(1).to_string()))
                    predicted_labels.append(predicted_label)
                    used_sentences.append(sentences[i])
                    result = {"ID": predicted_label, "Name": label_name, "Sentence": sentences[i],}
                    
                    df.loc[len(df)] = result
    df.drop_duplicates(inplace=True)
    return df

# total arguments
n = len(sys.argv)
print("Start analysing", n-1, "documents...")

for i in range(1, n):
    file_path = sys.argv[i]
    results = analyze_all_doc(file_path, ml_model_filenames)

    index = 0
    for char in file_path:
        if char == '.':
            break
        index+=1

    new_file_path = file_path[:index] + '_result.xlsx'
    # print(new_file_path)
    results.to_excel(new_file_path, index=False)
    print("Analysis progress: ", i, "/", n-1, sep="")



whodunit/whodunit-Hattie.py
#!/usr/bin/env python


from __future__ import print_function


__description__ = "Recognize an APT group from the techniques used"
__license__ = "GPL"
__VERSION__ = "1.0.3"
__uri__ = "https://gitlab.com/bontchev/whodunit"
__author__ = "Vesselin Bontchev"
__email__ = "vbontchev@yahoo.com"


from argparse import ArgumentParser
from heapq import nlargest, nsmallest
from os import access, R_OK
from os.path import isfile
from sys import argv, stderr, version_info
import pandas as pd #to convert excel


if version_info < (3, 6):
    print("Python 3.6 or higher required.", file=stderr)
    exit()

try:
    from enterpriseattack import Attack
except ImportError:
    print(
        'Could not import module "enterpriseattack"; try "pip install enterpriseattack".',
        file=stderr,
    )
    exit()


def get_options():
    parser = ArgumentParser(description=__description__)
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        version="%(prog)s version {}".format(__VERSION__),
    )
    parser.add_argument(
        "-n",
        "--numgroups",
        type=int,
        default=10,
        help="Number of closest groups to list (default: %(default)s)",
    )
    parser.add_argument(
        "-b", "--verbose", action="store_true", help="Verbose operation"
    )
    parser.add_argument(
        "-u", "--update", action="store_true", help="Update the MITRE ATT&CK data"
    )
    parser.add_argument(
        "-d",
        "--deprecated",
        action="store_true",
        help="Don't ignore the deprecated techniques",
    )
    parser.add_argument(
        "techniques",
        nargs="*" if "-u" in argv or "--update" in argv else "+",
        help="File containing a list of observed techniques",
    )
    return parser.parse_args()

# original code only reads .txt file
def read_report(report_file, all_techniques, verbose):
    print('Processing report file "{}"...'.format(report_file))
    with open(report_file, encoding="utf-8") as f:
        data = f.readlines()

    techniques = set()
    for line in data:
        line = line.strip()
        pos = line.find("#")
        if pos >= 0:
            line = line[:pos].strip()
        tids = line.split()
        for tid in tids:
            tid = tid.upper()
            if tid in all_techniques:
                techniques.add(tid)
            else:
                print('Unrecognized technique: "{}".'.format(tid), file=stderr)
                continue # ignore 'no technique number' error
                exit()

    if verbose:
        print("Observed techniques:")
        for technique in sorted(techniques):
            print(technique)
        print()

    return techniques

def get_groups_and_techniques(attack, techniques):
    def get_techniques(groupObj):
        # merge techniques with sub-techniques
        mergedTechsAndSubs = groupObj.techniques + groupObj.sub_techniques
        # save as a list of technique IDs
        techIDs = [t.id for t in mergedTechsAndSubs]
        # declare a list to extract parent techniques of sub-techniques e.g., grab 1566 from 1566.001
        parent_techniques = []

        # loop through sub_techniques for parent extraction
        for p in groupObj.sub_techniques:
            # split sub-techniques string on "."
            parent = p.id.split(".")
            # remove sub element
            del(parent[1])
            # append parent to declared
            parent_techniques += parent
        
        # append duplicates-removed list of parent techniques to techIDs
        techIDs += set(parent_techniques)
        return techIDs

    groups = []
    max_name = 0
    for group in attack.groups:
        the_group = {}
        the_group["name"] = group.name
        if len(group.name) > max_name:
            max_name = len(group.name)
        the_group["id"] = group.id
        the_techniques = set(get_techniques(group))
        the_group["techniques"] = the_techniques
        common_techniques = the_techniques.intersection(techniques)
        missing = techniques.difference(the_techniques)
        the_group["confidence"] = len(common_techniques) * 100.0 / len(techniques)
        the_group["number of missing techniques"] = len(missing)
        the_group["missing techniques"] = techniques.difference(the_techniques)
        groups.append(the_group)
    return groups, max_name


def match_groups(groups, num_groups, max_name, techniques):
    closest_groups = nlargest(num_groups, groups, key=lambda x: x["confidence"])
    # get groups with least number of unmatched techniques
    least_missing_techniques_groups = nsmallest(num_groups, groups, key=lambda x: x["number of missing techniques"])
    # for each APT group in groups database
    for APT in groups: 
        # keep a sum of positions of missing techniques
        missing_score = 0
        print("Position of missing technique in attack sequence of threat actor {}".format(APT["name"]))
        # for each missing technique of the APT group
        for technique in APT["missing techniques"]: 
            # convert techniques set to list then sort
            sorted_techniques = sorted(list(techniques))
            # find index of the missing technique in attack techniques list
            pos = sorted_techniques.index(technique) 
            print(
                "{:<{width}} @ {:<{pos_width}} / {}".format(
                technique, pos+1, len(techniques), width=10, pos_width=2))
            # add to sum
            missing_score += (pos+1)
        # work out the missing score - divide by the number of techniques
        missing_score /= len(techniques)
        # save missing score to threat actor key
        APT["missing score"] = missing_score
        print()

    # original
    if num_groups > 1:
        print("The {} APT groups".format(num_groups), end="")
    else:
        print("The APT group", end="")
    print(" most likely responsible:")
    for group in sorted(closest_groups, key=lambda x: (-x["confidence"], x["id"])):
        print(
            "{} ({:<{width}}): {:>6.2f}%".format(
                group["id"], group["name"], group["confidence"], width=max_name
            )
        )
    print()
    
    # check the number of unmatched techniques
    if num_groups > 1:
        print("The {} APT groups".format(num_groups), end="")
    else:
        print("The APT group", end="")
    print(" with the least missing techniques number:")
    for group in sorted(least_missing_techniques_groups, key=lambda x: (x["number of missing techniques"], x["id"])):
        if group["number of missing techniques"] == 0:
            print(
                "{} ({:<{width}}): {:>6d}".format(
                    group["id"], group["name"], group["number of missing techniques"], width=max_name
                )
            )
        else:
            print(
                "{} ({:<{width}}): {:>6d} ({})".format(
                    group["id"], group["name"], group["number of missing techniques"], group["missing techniques"], width=max_name
                )
            )
    print()

    # print missing score of unmatched techniques
    if num_groups > 1:
        print("Missing score of the {} APT groups:".format(num_groups), end="")
    else:
        print("Missing score of the APT group:", end="")
    print()
    for group in sorted(least_missing_techniques_groups, key=lambda x: (x["number of missing techniques"], x["id"])):
        print(
            "{} ({:<{width}}): {:.2f}".format(
                group["id"], group["name"], group["missing score"], width=max_name
            )
        )


def main():
    args = get_options()
    attack = Attack(include_deprecated=args.deprecated, update=args.update)
    if len(args.techniques) == 0:
        exit()
    multi_report = len(args.techniques) > 1
    all_techniques = [
        x.id for x in list(attack.techniques) + list(attack.sub_techniques)
    ]
    for report_file in args.techniques:
        if isfile(report_file) and access(report_file, R_OK):
            techniques = read_report(report_file, all_techniques, args.verbose)
            groups, max_name = get_groups_and_techniques(attack, techniques)
            match_groups(groups, args.numgroups, max_name, techniques)
            if multi_report:
                print()
        else:
            print('Could not open file "{}"'.format(report_file), file=stderr)


if __name__ == "__main__":
    main()
